presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java

```diff
@@ -296,16 +305,65 @@ public ConnectorSplitSource getSplits(
                 splitSchedulingContext.schedulerUsesHostAddresses(),
                 layout.isPartialAggregationsPushedDown());
 
-        HiveSplitSource splitSource = computeSplitSource(splitSchedulingContext, table, session, hiveSplitLoader);
+        HiveSplitSource splitSource = computeSplitSource(splitSchedulingContext, table, session, hiveSplitLoader, ratio);
         hiveSplitLoader.start(splitSource);
 
         return splitSource;
     }
 
+    private double getSplitScanRatio(
+            ConnectorSession session,
+            SchemaTableName tableName,
+            HiveTableLayoutHandle layout,
+            TransactionalMetadata metadata,
+            List<HivePartition> partitions)
+    {
+        double ratio = 1.0;
+        if (!isDynamicSplitSizesEnabled(session)) {
+            return ratio;
+        }
+        HiveTableHandle hiveTableHandle = new HiveTableHandle(tableName.getSchemaName(), tableName.getTableName());
```

Comment:
use the function mergeRequestedAndPredicateColumns().  It will also handle if the same struct columns appear in both sets with different subfields.